### ISSUE TITLE
Do we need to change anything about i18n of Preferences strings?

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/joist#772]
Displayed strings in the preferences dialog are currently translatable. The dialog is still rarely used but the dialog and its content are always imported so the translation utility presents all related strings.

@oliver-phet reported that a translator found preferences strings while translating density, but the Preferences dialog is not used in that sim so the translator wasn't able to see the strings used in context.

 The same thing happens with the Options dialog, but it may be more of an issue for Preferences because it has many more strings.

I would like to bring this up during dev meeting to see if anyone has thoughts about 1) is this a problem? 2) Is there a way to conditionally import the Preferences related components so these strings aren't available in the translation utility?

**Comment by chrisklus** [comment]
From 1/27/22 dev meeting:

@samreid: are we worried about bad translations? Could we make this testable in rosetta so they are more accurate?
@jbphet: that could be confusing if they see it for translations but then it doesn't exist in the real sim.

@zepumph: could an easy fix be just to put a warning message that those strings are for a specific thing that likely won't be in the sim? It feels like the priority is to reduce confusion, not necessarily stop people from translating things.

Consensus: @jessegreenberg is going to check in with a designer or KP to make sure this is a good plan and then make the changes in rosetta. Then he will pass it over to @jbphet.

**Comment by jessegreenberg** [comment]
This was discussed on slack, pasting the thread here:

<details>

A few questions:
Is there one sim that does have all of the preference menu strings in action? e.g. is JT a place where the translators could see the strings present and be able to translate them?
How do other "common strings" work now? When you translate a sim do you only see the common strings that appear in that sim, or do you see all the common code strings at the bottom, even if they are not in that specific sim?
One low-time alternative solution - if there is a sim that includes all the preference menu strings - would be for Oliver to email all the translators and just prompt them to translate that simulation. So they can do those translations in context.
But if the answer to (2) above is that "common strings" do not usually appear in a sim if they aren't visible in the sim, then how does that work for those common strings?





6:49
I added Liam and John - so we can all discuss.

John Blanco  6:56 PM
In answer to your question, strings from any files that are included in the build are shown in Rosetta.  It would be very difficult for Rosetta to know which strings are actually used.  And, as of recently, the Preferences Dialog is always included in the build, it's just not always shown. (edited)

1 reply
19 days agoView thread

Jesse Greenberg  6:59 PM
Thanks Kathy -  for the questions:
JT would be the one to reference, it has almost all of the strings. There are still a few strings presented that will not show up in that sim either yet related to "Gesture Controls" because we have some dialog content for that feature but haven't published a sim with that yet.
This is how other "common strings" work as well if the component is included in the build (like JB said). For example, the "Options" dialog strings are available when translating density, but that sim doesn't have an "Options" dialog. But it is more noticeable with Preferences because it has more content.
7:00
@jbphet is there a way to see if strings have been translated at all? If the "Gesture controls" strings have not been translated maybe we can remove them. This feature is on hold and there are no plans to publish with it any time soon. If those are removed then JT would be an example for all Preferences strings.

Kathy Perkins  7:03 PM
If Gesture controls aren't going to be published anytime soon, then I think it would also be OK to just wipe prior translations of these particular strings and remove it from rosetta.
7:04
I think having Oliver ask translators to do JT sounds correct - and not worrying about having a special message in rosetta.

Jesse Greenberg  7:06 PM
I like this! We would like these strings to be translated.
7:07
@jbphet @Liam is removing these strings something we can do?

Liam Mulhall  7:19 PM
We can edit translations by editing the string files in Babel and then prompting the build server to rebuild the translation. If there are lots of translations that have Gesture Controls strings, I think it will be difficult to remove them all. @jessegreenberg, what repo houses the Gesture Control strings?

Kathy Perkins  7:21 PM
Jesse - when you say gesture publication is delayed ... is it sort of on indefinite hold? any sense of the timeline?

John Blanco  7:24 PM
It's not too hard to remove previously translated strings.  I have a little script file to do it, which I should probably check in at some point.  That's why @Liam doesn't know about it.
:+1::+1::skin-tone-2:
3


Jesse Greenberg  10:49 AM
when you say gesture publication is delayed ... is it sort of on indefinite hold? any sense of the timeline?
I believe indefinite. It is not expected to get any time in 2022, nor is it very close to being publication ready.
10:49
 what repo houses the Gesture Control strings?
@Liam they are in joist.

Kathy Perkins  10:53 AM
OK - Then sounds reasonable to remove them from Rosetta if it is easy and won't create issues with existing sims and such.

Jesse Greenberg  10:56 AM
OK, thanks @kathy - Ill update the issue with these notes and then work with @jbphet and @Liam to remove the strings.
</details>

To summarize: Instead of a UI change to Rosetta, we are going to point translators to the example usage in john-travoltage when these questions come up. We are going to remove the strings from common code that are not used in a simulation yet because there is no example/usage of them and they should not be translated yet. Ill coordinate with @jbphet and @liam-mulhall to remove these strings. @jbphet has a script to remove some strings from all translations/strings files.

**Comment by jessegreenberg** [comment]
The strings to remove are:
"preferences.tabs.general.simulationSpecificSettings"
"preferences.tabs.input.title"
"preferences.tabs.input.gestureControls.title"
"preferences.tabs.input.gestureControls.description"

**Comment by zepumph** [comment]
So can we just remove based on the prefix: `preferences.`?

**Comment by jessegreenberg** [comment]
Not `preferences.`, but `preferences.tabs.input.`. Also the `preferences.tabs.general.simulationSpecificSettings`.

**Comment by zepumph** [comment]
I see, thanks for catching me up!

**Comment by jessegreenberg** [comment]
I copied the values of these string keys to their usages so simulation code is no longer referencing these keys. That means sims deployed starting now should not have these string show up in the translation interface and they should be ready to remove from joist-strings_en.json.

**Comment by jessegreenberg** [comment]
I connected with @jbphet and @liam-mulhall about removing these strings from joist-strings_en.json and they said to assign the issue to them. Thanks!

**Comment by jbphet** [comment]
I just went back through this issue, and I want to verify something before I pull the trigger and delete these strings.  First a little background: Translated common-code strings are not branched nor is a SHA recorded when a sim is published.  Instead, all translation builds get their common-code strings from the master version of the babel repo.  There are good reasons for this, but I won't get into them here.  The bottom line is that we shouldn't delete a string from the translated common-code string files if it is used in any existing published sim.  I just downloaded the live version of John Travoltage and checked whether the built HTML file contains these string keys, and it does.  So, before I get rid of these strings, I'd just like get a clear answer to this question: Are we sure that these strings are not actually used in this nor any other published sim?

It is indeed _possible_ for the strings to be used in the code but never presented to the user, but it's rare. This may well be the case here, and that's what I'm trying to confirm.

Assigning to @jessegreenberg to answer this, assign it back to me if the answer is yes and I'll finish this off.

**Comment by jessegreenberg** [comment]
>It is indeed possible for the strings to be used in the code but never presented to the user, but it's rare. This may well be the case here, and that's what I'm trying to confirm.

That is not the case. These strings were in the code and presented to the user. They are used in published sims. In https://github.com/phetsims/joist/commit/bd489a5e4b22489a3dbf7711166a42958f034f02 I removed them from the code so that future publications off of master do not show these strings.

We were hoping to remove these keys from existing translations: https://github.com/phetsims/joist/issues/772#issuecomment-1042109841. They would have been displayed to translators in published john-travoltage and gravity-force-lab-basics.

**Comment by jbphet** [comment]
@jessegreenberg and I just discussed this over Zoom, and there is an important point that we were able to clarify.  When @jessegreenberg said in the preceding comment:

> These strings were in the code and presented to the user. They are used in published sims.

...he meant that they were presented to the user in _Rosetta_, but they have never been visible to the user _in a simulation_.

This means that it is indeed safe to remove them from master, so I will proceed with that.

**Comment by jbphet** [comment]
This is done.  I used a script for this, and it is now checked in at https://github.com/phetsims/rosetta/blob/master/scripts/js/remove-string.js.

For some reason, when I used the script, it changed a bunch of unicode characters in the English joist strings file.  I didn't want to commit this, so I manually reverted these changes before doing the commit so only the removal of the strings was included.  I'm noting this for the next time we use the script.

@jessegreenberg - please look over the strings files and, if all looks good, this can be closed.

**Comment by jessegreenberg** [comment]
Excellent, thanks for doing this. I looked at the removed strings in https://github.com/phetsims/joist/commit/d044254a1d3fefb9da18b0848cf038c3ba9b4dff and the change is correct. CLosing.
